maildump/smtp.py

```diff
@@ -1,17 +1,115 @@
+            self.push('501 Syntax: EHLO hostname')
+            return
+        # See issue #21783 for a discussion of this behavior.
+        if self.seen_greeting:
+            self.push('503 Duplicate HELO/EHLO')
+            return
+        self._set_rset_state()
+        self.seen_greeting = arg
+        self.extended_smtp = True
+        self.push('250-%s' % self.fqdn)
+        if self._smtp_auth:
+            self.push('250-AUTH PLAIN')
+        if self.data_size_limit:
+            self.push('250-SIZE %s' % self.data_size_limit)
+            self.command_size_limits['MAIL'] += 26
+        if not self._decode_data:
+            self.push('250-8BITMIME')
+        if self.enable_SMTPUTF8:
+            self.push('250-SMTPUTF8')
+            self.command_size_limits['MAIL'] += 10
+        self.push('250 HELP')
+
+    def smtp_AUTH(self, arg):
+        print('auth:', arg, file=smtpd.DEBUGSTREAM)
+        if not self._smtp_auth:
+            self.push('501 Syntax: AUTH not enabled')
+            return
+
+        if not arg:
+            self.push('501 Syntax: AUTH TYPE base64(username:password)')
+            return
+
+        if not arg.lower().startswith('plain '):
+            self.push('501 Syntax: only PLAIN auth possible')
+            return
+
+        auth_type, auth_data = arg.split(None, 1)
+        try:
+            auth_data = base64.b64decode(auth_data.strip())
+        except TypeError:
+            self.push('535 5.7.8 Authentication credentials invalid')
+            return
+
+        if self.is_valid_user(auth_data):
+            self.push('235 Authentication successful')
+            self._authorized = True
+            return
+
+        self._authorized = False
+        self.push('535 5.7.8 Authentication credentials invalid')
+
+    def smtp_MAIL(self, arg):
+        if self._smtp_auth and not self._authorized:
+            self.push('530 5.7.0  Authentication required')
+            return
+        super(SMTPChannel, self).smtp_MAIL(arg)
+
+    def smtp_RCPT(self, arg):
+        if self._smtp_auth and not self._authorized:
+            self.push('530 5.7.0  Authentication required')
+            return
+        super(SMTPChannel, self).smtp_RCPT(arg)
+
+    def smtp_DATA(self, arg):
+        if self._smtp_auth and not self._authorized:
+            self.push('530 5.7.0  Authentication required')
+            return
+        super(SMTPChannel, self).smtp_DATA(arg)
+
+
 class SMTPServer(smtpd.SMTPServer, object):
-    def __init__(self, listener, handler):
+    def __init__(self, listener, handler, smtp_auth):
         super(SMTPServer, self).__init__(listener, None)
         self._handler = handler
+        self._smtp_auth = smtp_auth
+
+    def handle_accept(self):
+        pair = self.accept()
+        if pair is not None:
+            conn, addr = pair
+            print('Incoming connection from %s' % repr(addr), file=smtpd.DEBUGSTREAM)
+            channel = SMTPChannel(self, conn, addr, self._smtp_auth)
```

Comment:
Don't override this; set the `channel_class` attribute instead. Ideally, only do that if SMTP AUTH is enabled in order to avoid using the custom channel implementation in case it's not used.

Reply:
This channel needs one param more. Anyway, I've changed the way how it's handled a little bit.